client/lib/logmein/index.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Internal Dependencies
+ */
+import { isEnabled } from '@automattic/calypso-config';
+import wpcom from 'calypso/lib/wp';
+
+/**
+ * Append logmein=1 query parameter to mapped domain urls we want the user to be logged in against.
+ */
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
+	let newurl: URL;
+
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	let allow = logmeinAllowedUrls( sites );
+	// Ignore urls not in the allow list
+	allow = allow.map( ( allowed: string ) => new URL( allowed, INVALID_URL ).hostname );
+	if ( allow.indexOf( newurl.hostname ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// we're already logged into wordpress.com so we can take a shortcut
+	if ( isWPComLoggedIn ) {
+		newurl.searchParams.set( 'logmein', 'direct' );
+	} else {
+		newurl.searchParams.set( 'logmein', '1' );
+	}
+	return newurl.toString();
+}
+
+export function logmeinAllowedUrls( sites: any ): string[] {
+	return sites
+		.map( ( site: any ) =>
+			! site.is_vip &&
+			! site.jetpack &&
+			! site.options.is_automated_transfer &&
+			! site.options.is_domain_only &&
+			! site.options.is_redirect &&
+			! site.options.is_wpcom_atomic &&
+			! site.options.is_wpcom_store &&
+			! site.options.is_wpforteams_site &&
+			site.options.is_mapped_domain
+				? new URL( site.URL ).host
+				: false
+		)
+		.filter( Boolean );
+}
+
+export function appendLogmein( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', '1' );
+	return url;
+}
+export function appendLogmeinDirect( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', 'direct' );
+	return url;
+}
+
+let allowedSites: string[] = [];
+
+export function attachLogmein( isWPComLoggedIn = false ): void {
+	if ( ! isEnabled( 'logmein' ) ) {
+		return;
+	}
+
+	if ( isWPComLoggedIn ) {
+		wpcom
+			.me()
+			.sites()
+			.then( ( sites: any ) => {
+				allowedSites = logmeinAllowedUrls( sites.sites );
+				document.addEventListener( 'click', logmeinOnClick, false );
+			} );
+	}
+}
+
+function logmeinOnClick( event: MouseEvent ) {
+	const link = ( event.target as HTMLElement ).closest( 'a' );
```

Comment:
I'm wondering what the performance implications of adding a tree traversal to every single mouse click in the app will be. Perhaps the nature of mouse clicks mean this'll be fine (it's not like dragging or typing, the user can probably tolerate a few milliseconds delay for a click). Just thought it was worth bringing up since this is very global.

Reply:
Definitely worth bringing up.

The traversal should be linear time so not too concerning. I'm guessing `closest` [looks like this](https://github.com/Automattic/wp-calypso/pull/53443/commits/1585505622ab70fdd7db602aa7ea16e8a2500dfb#diff-d025c6a21229fe69836222916356ee8a2c395861252a82ff88e97b9facc70a4bL95) chaining up `.parent`. I would like to back this up if we can though the globalness of this is what concerns me the most.

Lets discuss this [above here](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647994782) as watching a mutation observer for links and appending preemptively might be a better option.

Reply:
> I'm wondering what the performance implications of adding a tree traversal to every single mouse click in the app will be.

The performance impact should be negligible, because this tree traversal already happens many many times on every event, and adding one more is unlikely to be a concern.

React itself adds a global `click` listener to the `document` or the React root element. When a click happens on e.g., a `<button>`, React is traversing the DOM tree from the `<button>` element up to the React root and dispatches synthetic events on the way.

The `page.js` router installs a global `click` handler where it intercepts all clicks on `<a href>` links and handles them using `page.js` code instead of letting the browser do a native navigation with full reload. That means that a click handler almost identical to the logmein one is already active in Calypso.

The browser itself does a DOM traversal from the root `window` element to the target one (capturing) and back to root again (bubbling), finding and invoking event handlers on the way.

Browser extensions can install their own intercepting event handlers. For example, my Firefox Devtools tell me that the 1Password extension is watching all `click`s on my `<button>` elements, using a `closest( 'button' )`-like traversal.

Reply:
Too much TIL in this PR. :bow: 

Good to know this is a rather well accepted/safe approach to attach click events in a generalized way.

Reply:
Great points 👍